src/main/java/org/cancogenvirusseq/muse/components/SongScoreClient.java

```diff
@@ -0,0 +1,216 @@
+package org.cancogenvirusseq.muse.components;
+
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
+import java.util.Objects;
+import javax.annotation.PostConstruct;
+import lombok.Builder;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+import org.cancogenvirusseq.muse.model.song_score.AnalysisFileResponse;
+import org.cancogenvirusseq.muse.model.song_score.ScoreFileSpec;
+import org.cancogenvirusseq.muse.model.song_score.SubmitResponse;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.http.MediaType;
+import org.springframework.stereotype.Component;
+import org.springframework.web.reactive.function.BodyInserters;
+import org.springframework.web.reactive.function.client.WebClient;
+import reactor.core.publisher.Mono;
+import reactor.util.function.Tuple2;
+import reactor.util.function.Tuples;
+
+@Slf4j
+@Component
+public class SongScoreClient {
+
+  @Value("${songScoreClient.songRootUrl}")
+  String songRootUrl;
+
+  @Value("${songScoreClient.scoreRootUrl}")
+  String scoreRootUrl;
+
+  @Value("${songScoreClient.systemApiToken}")
+  String systemApiToken;
+
+  @PostConstruct
+  public void init() {
+    log.info("Initialized song score client.");
+    log.info("songRootUrl - " + songRootUrl);
+    log.info("scoreRootUrl - " + scoreRootUrl);
+  }
+
+  // main function executing the song and dance
+  public Mono<String> submitAndUpload(
+      String studyId, String payload, String fileContent, String fileMd5) {
+    return submitPayload(studyId, payload, fileMd5)
+        .flatMap(this::getFileSpecFromSong)
+        .flatMap(this::initScoreUpload)
+        .flatMap(dto -> uploadToS3(dto, fileContent))
+        .flatMap(this::finalizeScoreUpload)
+        .flatMap(this::publishAnalysis);
```

Comment:
Side note, I think we should seriously consider a new reactor/webflux song-score client cause I think it would be much more cleaner... just my thoughts